waku/waku_api/message_cache.nim

```diff
@@ -15,68 +15,286 @@ import
+  contentIndex: seq[(int, int)] # Content topic index mapped to message index
 
-type MessageCacheResult*[T] = Result[T, cstring]
+func `$`*(self: MessageCache): string =
+  "Messages: " & $self.messages.len &
+  " \nPubsubTopics: " & $self.pubsubTopics &
+  " \nContentTopics: " & $self.contentTopics &
+  " \nPubsubIndex: " & $self.pubsubIndex &
+  " \nContentIndex: " & $self.contentIndex
 
-type MessageCache*[K] = ref object
-  capacity: uint
-  table: Table[K, seq[WakuMessage]]
-
-func init*[K](T: type MessageCache[K], capacity=DefaultMessageCacheCapacity): T =
-  MessageCache[K](
-    capacity: capacity,
-    table: initTable[K, seq[WakuMessage]]()
+func init*(T: type MessageCache, capacity=DefaultMessageCacheCapacity): T =
+  MessageCache(
+    capacity: capacity
   )
 
+proc messagesCount*(self: MessageCache): int =
+  self.messages.len
 
-proc isSubscribed*[K](t: MessageCache[K], topic: K): bool =
-  t.table.hasKey(topic)
+proc pubsubSearch(self: MessageCache, pubsubTopic: PubsubTopic): Option[int] =
+  # Return some with the index if found none otherwise.
+  
+  for i, topic in self.pubsubTopics:
+    if topic == pubsubTopic:
+      return some(i)
 
-proc subscribe*[K](t: MessageCache[K], topic: K) =
-  if t.isSubscribed(topic):
-    return
-  t.table[topic] = @[]
+  return none(int)
 
-proc unsubscribe*[K](t: MessageCache[K], topic: K) =
-  if not t.isSubscribed(topic):
-    return
-  t.table.del(topic)
+proc contentSearch(self: MessageCache, contentTopic: ContentTopic): Option[int] =
+  # Return some with the index if found none otherwise.
+  
+  for i, topic in self.contentTopics:
+    if topic == contentTopic:
+      return some(i)
+
+  return none(int)
+
+proc isPubsubSubscribed*(self: MessageCache, pubsubTopic: PubsubTopic): bool =
+  self.pubsubSearch(pubsubTopic).isSome()
+
+proc isContentSubscribed*(self: MessageCache, contentTopic: ContentTopic): bool =
+  self.contentSearch(contentTopic).isSome()
+
+proc pubsubSubscribe*(self: MessageCache, pubsubTopic: PubsubTopic) =
+  if self.pubsubSearch(pubsubTopic).isNone():
+    self.pubsubTopics.add(pubsubTopic)
+
+proc contentSubscribe*(self: MessageCache, contentTopic: ContentTopic) =
+  if self.contentSearch(contentTopic).isNone():
+    self.contentTopics.insert(contentTopic)
+
+proc removeMessage(self: MessageCache, idx: int) =
+  let lastIndex = self.messages.high
+  self.messages.del(idx)
+  
+  # update indexes
+  var j = self.pubsubIndex.high
+  while j > -1:
+    let (pId, mId) = self.pubsubIndex[j]
+
+    if mId == idx:
+      self.pubsubIndex.del(j)
+    elif mId == lastIndex:
+      self.pubsubIndex[j] = (pId, idx)
+
+    dec(j)
+
+  j = self.contentIndex.high
+  while j > -1:
+    let (cId, mId) = self.contentIndex[j]
+
+    if mId == idx:
+      self.contentIndex.del(j)
+    elif mId == lastIndex:
+      self.contentIndex[j] = (cId, idx)
+
+    dec(j)
+
+  if self.writeIndex > self.messages.high:
+    self.writeIndex = self.messages.high
```

Comment:
If you delete a message mid-array, the indexes of the messages after it get decreased by one and then `mId` in `pubsubIndex` and `contentIndex` can point to different messages than what they initially intended.

Is that taken care of somewhere? or am I missing something?

I feel that the logic should be much simpler and less bug-prone with a different selection of data structures.
I think that using a hash table instead of a seq, and using messageIds as identifiers instead of their indexes should simplify a lot the logic and avoid the problem of moving indexes

Let me know what you think :):)

Reply:
I agree it's a bit over engineered. :D I just normalized the data and ran with it...... was fun to write though. :P

In this case the `del` is a swap to last index then a decrement of the seq length. O(1) but doesn't preserve ordering.
https://nim-lang.org/docs/system.html#del,seq[T],Natural

Reply:
Oh makes sense! Great then :) Thank you!

Reply:
Now that I think, if ordering is not preserved then when we add a message when the cache is full we may have the unintended consequence of removing a message that is not necessarily the oldest one.

Is that ok?

Reply:
:thinking: we could check the message timestamp instead?